core/chain_pow_test.go

```diff
@@ -0,0 +1,232 @@
+			var results <-chan nonceCheckResult
+
+			switch {
+			case full && valid:
+				_, results = verifyBlockNonces(FakePow{}, blocks)
+			case full && !valid:
+				_, results = verifyBlockNonces(failPow{uint64(len(blocks) - 1)}, blocks)
+			case !full && valid:
+				_, results = verifyHeaderNonces(FakePow{}, headers)
+			case !full && !valid:
+				_, results = verifyHeaderNonces(failPow{uint64(len(headers) - 1)}, headers)
+			}
+			// Wait for all the verification results
+			checks := make(map[int]bool)
+			for k := 0; k < len(blocks); k++ {
+				select {
+				case result := <-results:
+					if _, ok := checks[result.index]; ok {
+						t.Fatalf("test %d.%d.%d: duplicate results for %d", i, j, k, result.index)
+					}
+					if result.index < 0 || result.index >= len(blocks) {
+						t.Fatalf("test %d.%d.%d: result %d out of bounds [%d, %d]", i, j, k, result.index, 0, len(blocks)-1)
+					}
+					checks[result.index] = result.valid
+
+				case <-time.After(time.Second):
+					t.Fatalf("test %d.%d.%d: verification timeout", i, j, k)
+				}
+			}
+			// Check nonce check validity
+			for k := 0; k < len(blocks); k++ {
+				want := valid || (k != len(blocks)-2)
+				if checks[k] != want {
+					t.Errorf("test %d.%d.%d: validity mismatch: have %v, want %v", i, j, k, checks[k], want)
+				}
+			}
+			// Make sure no more data is returned
+			select {
+			case result := <-results:
+				t.Fatalf("test %d.%d: unexpected result returned: %v", i, j, result)
+			case <-time.After(25 * time.Millisecond):
+			}
+		}
+	}
+}
+
+// Tests that aborting a POW validation indeed prevents further checks from being
+// run, as well as checks that no left-over goroutines are leaked.
+func TestPowConcurrentAbortion2(t *testing.T)  { testPowConcurrentAbortion(t, 2) }
+func TestPowConcurrentAbortion8(t *testing.T)  { testPowConcurrentAbortion(t, 8) }
+func TestPowConcurrentAbortion32(t *testing.T) { testPowConcurrentAbortion(t, 32) }
+
+func testPowConcurrentAbortion(t *testing.T, threads int) {
+	// Create a simple chain to verify
+	var (
+		testdb, _ = ethdb.NewMemDatabase()
+		genesis   = GenesisBlockForTesting(testdb, common.Address{}, new(big.Int))
+		blocks    = GenerateChain(genesis, testdb, 1024, nil)
+	)
+	headers := make([]*types.Header, len(blocks))
+	for i, block := range blocks {
+		headers[i] = block.Header()
+	}
+	// Set the number of threads to verify on
+	old := runtime.GOMAXPROCS(threads)
+	defer runtime.GOMAXPROCS(old)
+
+	// Run the POW checker for blocks one-by-one, checking for both valid and invalid nonces
+	for i, full := range []bool{true, false} {
+		var abort chan<- struct{}
+		var results <-chan nonceCheckResult
+
+		// Start the verifications and immediately abort
+		if full {
+			abort, results = verifyBlockNonces(delayedPow{time.Millisecond}, blocks)
+		} else {
+			abort, results = verifyHeaderNonces(delayedPow{time.Millisecond}, headers)
+		}
+		close(abort)
+
+		// Deplete the results channel
+		verified := make(map[int]struct{})
```

Comment:
We don't write code for people "new to Go". We write proper Go code for people (at the very least this is our aim).

An empty `struct{}` is preferable because it's being used to do inclusion checks; we're not using the actual data (difference between `true` and `false`) within the map it self. Secondly (altho negligible) an empty `struct{}` cost nothing in terms of memory (iirc) and a `bool`ean does.